os/fs/smartfs/smartfs_utils.c

```diff
@@ -1268,6 +1268,127 @@ int smartfs_createentry(struct smartfs_mountpt_s *fs, uint16_t parentdirsector,
+ *              writing the changed entry to MTD flash.
+ *
+ ****************************************************************************/
+
+int smartfs_replaceentry(struct smartfs_mountpt_s *fs, struct smartfs_entry_s oldentry, const char *filename, uint16_t type, struct smartfs_entry_s *direntry, FAR struct smartfs_ofile_s *sf)
+{
+	int ret;
+	struct smart_read_write_s readwrite;
+	uint16_t nextsector;
+	uint16_t entrysize;
+#ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
+	struct smartfs_chain_header_s *chainheader;
+#endif
+	uint8_t chainheader_type;
+
+	entrysize = sizeof(struct smartfs_entry_header_s) + fs->fs_llformat.namesize;
+
+	/*Read the entire logical sector containing the old entry into fs_rwbuffer*/
+	readwrite.logsector = oldentry.dsector;
+	readwrite.count = fs->fs_llformat.availbytes;
+	readwrite.offset = 0;
+	readwrite.buffer = (uint8_t *)fs->fs_rwbuffer;
+	ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
+	if (ret < 0) {
+		fdbg("Error %d reading sector %d data\n", ret, oldentry.dsector);
+		goto errout;
+	}
+
+	if (oldentry.firstsector == 0xFFFF) {
+		/* Allocate a new sector for the file / dir */
+
+		ret = FS_IOCTL(fs, BIOC_ALLOCSECT, 0xFFFF);
+		if (ret < 0) {
+			goto errout;
+		}
+
+		nextsector = (uint16_t)ret;
+
+		/* Set the newly allocated sector's type (file or dir) */
+
+#ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
+		if (sf) {
+			/* Using sector buffer and we have an open file context.  Just update
+			 * the sector buffer in the open file context.
+			 */
+
+			memset(sf->buffer, CONFIG_SMARTFS_ERASEDSTATE, fs->fs_llformat.availbytes);
+			chainheader = (struct smartfs_chain_header_s *)sf->buffer;
+			chainheader->type = SMARTFS_SECTOR_TYPE_FILE;
+			sf->bflags = SMARTFS_BFLAG_DIRTY | SMARTFS_BFLAG_NEWALLOC;
+		} else
+#endif
+		{
+			if ((type & SMARTFS_DIRENT_TYPE) == SMARTFS_DIRENT_TYPE_DIR) {
+				chainheader_type = SMARTFS_SECTOR_TYPE_DIR;
+			} else {
+				chainheader_type = SMARTFS_SECTOR_TYPE_FILE;
+			}
+
+			readwrite.count = 1;
+			readwrite.offset = offsetof(struct smartfs_chain_header_s, type);
+			readwrite.buffer = (uint8_t *)&chainheader_type;
+			readwrite.logsector = nextsector;
+			ret = FS_IOCTL(fs, BIOC_WRITESECT, (unsigned long)&readwrite);
+			if (ret < 0) {
+				fdbg("Error %d setting new sector type for sector %d\n", ret, nextsector);
+				goto errout;
+			}
+		}
+	} else {
+		/* Use the provided sector number */
+
+		nextsector = oldentry.firstsector;
+	}
+
+	/* Now write the new entry to the parent directory sector */
+	readwrite.logsector = oldentry.dsector;
+	readwrite.offset = oldentry.doffset;
+	readwrite.count = entrysize;
+	readwrite.buffer = (uint8_t *)&fs->fs_rwbuffer[oldentry.doffset];
+	ret = FS_IOCTL(fs, BIOC_WRITESECT, (unsigned long) &readwrite);
+	if (ret < 0) {
+		fdbg("failed to write new entry to parent directory psector : %d\n", oldentry.dsector);
+		goto errout;
+	}
+
+	/* Now fill in the entry */
```

Comment:
please check this very carefully.
It assume that only rename case.
If it is open, direntry->dsector should be newly allocated sector in above .